pil-analyzer/src/traits_resolver.rs

```diff
@@ -9,54 +9,44 @@ use std::{collections::HashMap, sync::Arc};
 
 use crate::type_unifier::Unifier;
 
-/// Resolves trait implementations for all definitions and identities.
-/// Checks references for calls to trait implementations
-/// and unifies them using their type args.
-///
-/// Returns a nested map of trait name to a map of type args to the expression.
-pub fn traits_resolution(
-    references: Vec<PolynomialReference>,
-    trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
-) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
-    TraitsResolver::new().resolve_traits(references, trait_impls)
-}
-
 type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
 
-struct TraitsResolver {}
+pub struct TraitsResolver {
+    trait_impls: HashMap<String, Vec<TraitImplementation<Expression>>>,
+    solved_impls: HashMap<String, HashMap<Vec<Type>, Arc<Expression>>>,
+}
 
 impl TraitsResolver {
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>) -> Self {
+        Self {
+            trait_impls: trait_impls.clone(),
+            solved_impls: HashMap::new(),
+        }
     }
 
-    pub fn resolve_traits(
-        &self,
-        references: Vec<PolynomialReference>,
-        trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
-    ) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
-        let mut solved_impls = HashMap::new();
-
-        for ref_poly in references {
-            if let Some(solved) = self.resolve_trait_function_reference(&ref_poly, trait_impls) {
-                let (key, type_args) = solved.0;
-                solved_impls
+    pub fn resolve_trait(&mut self, ref_poly: &PolynomialReference) -> Result<(), String> {
+        match self.resolve_trait_function_reference(ref_poly) {
+            Some(((key, type_args), expr)) => {
+                self.solved_impls
                     .entry(key)
-                    .or_insert_with(HashMap::new)
-                    .insert(type_args, solved.1);
+                    .or_default()
+                    .insert(type_args, expr);
```

Comment:
It would be nice if we could avoid running the unifier and all that in case we already have an entry in the hash maps here.